server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -697,6 +695,49 @@ protected int cleanupExpiredPoolsImpl() {
         return entitlementsToRegen;
     }
 
+    Set<Pool> revokeEntitlementsFromOverflowingPools(List<Pool> pools) {
+        Collection<Pool> overflowing = new ArrayList<>();
+        for (Pool pool : pools) {
+            if (pool.isOverflowing()) {
+                overflowing.add(pool);
+            }
+        }
+        if (overflowing.isEmpty()) {
+            return null;
+        }
+        List<Entitlement> entitlementsToRevoke = new ArrayList<>();
+        overflowing = poolCurator.lockAndLoad(overflowing);
+
+        List<Entitlement> overFlowingEnts = this.poolCurator.retrieveOrderedEntitlementsOf(overflowing);
+        Map<String, List<Entitlement>> entMap = new HashMap<>();
+        for (Entitlement entitlement : overFlowingEnts) {
+            List<Entitlement> poolEnts = entMap.get(entitlement.getPool().getId());
+            if (poolEnts == null) {
+                poolEnts = new ArrayList<Entitlement>();
+                entMap.put(entitlement.getPool().getId(), poolEnts);
+            }
+            poolEnts.add(entitlement);
+        }
+
+        for (Pool pool : overflowing) {
+            // we then start revoking the existing entitlements
+            List<Entitlement> entitlements = entMap.get(pool.getId());
+            long newConsumed = pool.getConsumed();
+
+            for (Entitlement ent : entitlements) {
+                if (newConsumed > pool.getQuantity()) {
```

Comment:
As written, this is a basically non-deterministic for revocations, right?  Would it make sense to apply some sort of heuristic to keep things consistent?  For example, sort the entitlements by quantity ascending and revoke the smallest entitlements first.